pageserver/src/tenant/layer_map.rs

```diff
@@ -200,45 +200,61 @@ where
     /// NOTE: This only searches the 'historic' layers, *not* the
     /// 'open' and 'frozen' layers!
     ///
-    pub fn search(&self, key: Key, end_lsn: Lsn) -> Option<SearchResult<L>> {
-        let version = self.historic.get().unwrap().get_version(end_lsn.0 - 1)?;
-        let latest_delta = version.delta_coverage.query(key.to_i128());
-        let latest_image = version.image_coverage.query(key.to_i128());
-
-        match (latest_delta, latest_image) {
-            (None, None) => None,
-            (None, Some(image)) => {
-                let lsn_floor = image.get_lsn_range().start;
-                Some(SearchResult {
-                    layer: image,
-                    lsn_floor,
-                })
-            }
-            (Some(delta), None) => {
-                let lsn_floor = delta.get_lsn_range().start;
-                Some(SearchResult {
-                    layer: delta,
-                    lsn_floor,
-                })
-            }
-            (Some(delta), Some(image)) => {
-                let img_lsn = image.get_lsn_range().start;
-                let image_is_newer = image.get_lsn_range().end >= delta.get_lsn_range().end;
-                let image_exact_match = img_lsn + 1 == end_lsn;
-                if image_is_newer || image_exact_match {
+    pub fn search(&self, key: Key, mut end_lsn: Lsn) -> Option<SearchResult<L>> {
+        loop {
```

Comment:
It's more efficient to change the insert method instead of search (insert the occupied ranges separately). Maybe leave a TODO if that's not planned for this PR.

Reply:
Will it be ok to have multiple references to the same layers?
Can  it cause returning the same layer twice in iteration through layers?

Reply:
When iterating the coverage it's good to return the same layer multiple times. But `iter_historic` should return each layer only once.